### ISSUE TITLE
Custom PyTorch BSDF with wrap_ad crashes

### ISSUE THREAD

**Issue by msqrt** [mitsuba-renderer/mitsuba3#945]
## Summary

Calling PyTorch code in functions wrapped within `wrap_ad` from a custom BSDF class crashes during rendering.

## System configuration
  OS: Windows-10
  CPU: Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
  GPU: NVIDIA TITAN RTX
  Python: 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)]
  NVidia driver: 536.23
  CUDA: 11.7.99
  LLVM: 13.-1.-1

  Dr.Jit: 0.4.3
  Mitsuba: 3.4.0
     Is custom build? False
     Compiled with: MSVC 19.37.32822.0
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb

## Description

I'm trying to implement a BSDF that runs PyTorch models for sampling, evaluation and pdf evaluation. I'd like to be able to propagate gradients through these, so I've separated the PyTorch calls into free functions wrapped with `dr.wrap_ad` that the BSDF class then calls. The rendering operation crashes with output `jit_var_eval(): variable r1135 remains dirty after evaluation!` whenever my custom BSDF is present in a scene -- the number in the error varies between runs, but otherwise the behavior is consistent. The suggested solution to similar issue #452 was to use `dr.set_flag(dr.JitFlag.VCallRecord, False)`, but this only silences the error output. It's also different, since in that issue the tensors were manually transferred between torch and drjit; as far as I understand, this is no longer encouraged / doesn't support automatic differentiation.

## Steps to reproduce

The following code produces the issue:

```
import torch

import drjit as dr
#dr.set_flag(dr.JitFlag.VCallRecord, False)

import mitsuba as mi
mi.set_variant('cuda_ad_rgb')

@dr.wrap_ad(source='drjit', target='torch')
def surface_pdf(surface, wi, n, wo, uv, duv_dx, duv_dy):
    return torch.ones_like(wi[..., 0].view(-1))

@dr.wrap_ad(source='drjit', target='torch')
def surface_eval(surface, wi, n, wo, uv, duv_dx, duv_dy):
    return wi

@dr.wrap_ad(source='drjit', target='torch')
def surface_sample(surface, wi, n, uv, duv_dx, duv_dy):
    return wi, torch.ones_like(wi[..., 0].view(-1)), wi

class Surface(mi.BSDF):
    def __init__(self, props):
        mi.BSDF.__init__(self, props)
        self.surface_type = props['surface_type']
        self.material_name = props['material_name']
        self.surface = None # load based on above values

        # Set the BSDF flags
        reflection_flags   = mi.BSDFFlags.GlossyReflection | mi.BSDFFlags.FrontSide | mi.BSDFFlags.Glossy
        self.m_components  = [reflection_flags]
        self.m_flags = reflection_flags

    def sample(self, ctx, si, sample1, sample2, active):
        bs = mi.BSDFSample3f()
        bs.eta = 1.
        bs.sampled_component = 0
        bs.sampled_type = mi.BSDFFlags.GlossyReflection
        wo, pdf, value = surface_sample(self.surface, *[mi.TensorXf(x) for x in (si.wi, si.n, si.uv, si.duv_dx, si.duv_dy)])
        bs.wo = dr.unravel(mi.Vector3f, wo)
        bs.pdf = dr.ravel(pdf)
        return bs, dr.unravel(mi.Vector3f, value)

    def eval(self, ctx, si, wo, active):
        return dr.unravel(mi.Vector3f, surface_eval(self.surface, *[mi.TensorXf(x) for x in (si.wi, si.n, wo, si.uv, si.duv_dx, si.duv_dy)]))
    
    def pdf(self, ctx, si, wo, active):
        return dr.ravel(surface_pdf(self.surface, *[mi.TensorXf(x) for x in (si.wi, si.n, wo, si.uv, si.duv_dx, si.duv_dy)]))


mi.register_bsdf("surface_wrapper", lambda props: Surface(props))

scene = mi.load_dict({
    'type': 'scene',
    'integrator': { 'type': 'path', },
    'sensor': {
        'type': 'perspective',
        'film': { 'type': 'hdrfilm', 'width': 512, 'height': 512, }
    },
    'bsdf': { 'type':'surface_wrapper', 'surface_type':'', 'material_name':'' }
})

import matplotlib.pyplot as plt
plt.imshow(mi.render(scene))
plt.show()
```

**Comment by rtabbara** [comment]
Hi @msqrt,

I think the problem you're encountering now isn't specific to `wrap_ad`, but rather a consequence of disabling `VCallRecord` and invoking a render call that by default uses an implicit recorded (symbolic) loop. In particular, for JIT-variants, a path integrator will trace the iterative path-sampling as a recorded loop that would otherwise result in multiple kernel launches at each iteration.

Disabling `VCallRecord` splits up a virtual function call into multiple kernel dispatches which requires knowledge of how to perform that split (e.g. for a set of ray interactions launch a kernel of size x for BSDF A, launch kernel of size y for BSDF B etc.). But if you have loop-dependent variables (e.g. ray) that are required to determine that split, that defeats the purpose of having a recorded loop to begin with as you would need your variables to be evaluated at each iteration. 

So in short, I think the solution here is that you'll also need to set `dr.set_flag(dr.JitFlag.LoopRecord, False)`. With #452, I don't think this was additionally required because they were using a direct illumination integrator so no recorded loop was being used in this instance.

**Comment by msqrt** [comment]
Hi!

Ah alright, that does make sense -- and setting that flag to false indeed solves the issue. Thanks a bunch, both for the explanation and the fix!

I'll close with a happy little rendering from the fixed BSDF.
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/1880715/5c916e53-c5f9-42e0-81c3-3e9599b803a1)
